prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,152 @@ def check_for_location_conflicts(substations):
         )
 
 
+def map_lines_to_substations_using_coords(substations, lines, **kwargs):
+    """Map lines to substations using coordinates
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param \\*\\*kwargs: optional arguments.
+    :return: (*pandas.DataFrame*) -- indices: lines id, columns: lines
+        endpoint coordinates (*FROM* and *TO*) and id of substations found
+        at these coordinates (*FROM_SUB* and *TO_SUB*)
+    """
+    rounding = 3 if "rounding" not in kwargs else kwargs["rounding"]
+    subcoord2subid = (
+        substations.round(rounding).groupby(["LONGITUDE", "LATITUDE"]).groups
+    )
+    print(
+        f"initial number of substations: {substations.shape[0]}\n"
+        f"substations with same location after rounding: "
+        f"{substations.shape[0] - len(subcoord2subid)}\n"
+    )
+
+    lines_coord_from = lines["COORDINATES"].map(
+        lambda x: (round(x[0][0], rounding), round(x[0][1], rounding))
+    )
+    lines_coord_to = lines["COORDINATES"].map(
+        lambda x: (round(x[-1][0], rounding), round(x[-1][1], rounding))
+    )
+    line2sub = lines_coord_from.to_frame(name="FROM").join(
+        lines_coord_to.to_frame(name="TO")
+    )
```

Comment:
Maybe we could do following here:
```python
lines_coord = lines.COORDINATES.map(lambda x: list(np.round([x[0],x[-1]], rounding)))
line2sub = pd.DataFrame(line_coord.to_list(), columns=["FROM", "TO"], index=line_coord.index)
```

Reply:
The problem is that the above gives me the lat/lon in `line2sub` as `list`. And then, I have the following error:
```
  File "/Users/brdo/CEM/PreREISE/prereise/gather/griddata/hifld/data_process/transmission.py", line 70, in <lambda>
    lambda x: list(subcoord2subid[x]) if x in subcoord2subid else np.NaN
TypeError: unhashable type: 'numpy.ndarray'
```

Reply:
Yeah, we need tuple here. Googling a bit and I found a way to convert all elements into `tuple` using: `line2sub.applymap(tuple)`

Reply:
Done